storage/uploader.go

```diff
@@ -0,0 +1,353 @@
+}
+
+// NewStorageUploader creates a new uploader instance
+func NewStorageUploader(config UploaderConfig) (*StorageUploader, error) {
+	if config.StorageURL == "" {
+		return nil, errors.New("storage URL is required")
+	}
+	if config.Wallet == nil {
+		return nil, errors.New("wallet is required for authentication")
+	}
+
+	// Create auth fetch client
+	authClient := authhttp.New(config.Wallet, nil, nil)
+
+	return &StorageUploader{
+		baseURL:   config.StorageURL,
+		authFetch: authClient,
+	}, nil
+}
+
+// uploadInfo is used to parse the response from the /upload endpoint
+type uploadInfo struct {
+	Status          string            `json:"status"`
+	UploadURL       string            `json:"uploadURL"`
+	RequiredHeaders map[string]string `json:"requiredHeaders"`
+	Amount          *int64            `json:"amount,omitempty"`
+}
+
+// getUploadInfo requests information from the server to upload a file
+func (u *StorageUploader) getUploadInfo(ctx context.Context, fileSize int, retentionPeriod int) (*uploadInfo, error) {
+	url := fmt.Sprintf("%s/upload", u.baseURL)
+	requestBody := map[string]interface{}{
+		"fileSize":        fileSize,
+		"retentionPeriod": retentionPeriod,
+	}
+
+	requestJSON, err := json.Marshal(requestBody)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal request body: %w", err)
+	}
+
+	reqOpts := &authhttp.SimplifiedFetchRequestOptions{
+		Method: "POST",
+		Headers: map[string]string{
+			"Content-Type": "application/json",
+		},
+		Body: requestJSON,
+	}
+
+	resp, err := u.authFetch.Fetch(ctx, url, reqOpts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get upload info: %w", err)
+	}
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("upload info request failed: HTTP %d", resp.StatusCode)
+	}
+
+	var info uploadInfo
+	if err := json.NewDecoder(resp.Body).Decode(&info); err != nil {
+		return nil, fmt.Errorf("failed to decode upload info response: %w", err)
+	}
+
+	if info.Status == "error" {
+		return nil, errors.New("upload route returned an error")
+	}
+
+	return &info, nil
+}
+
+// uploadFile performs the file upload to the presigned URL
+func (u *StorageUploader) uploadFile(ctx context.Context, uploadURL string, file UploadableFile, requiredHeaders map[string]string) (UploadFileResult, error) {
+	// Create HTTP client
+	client := &http.Client{}
+
+	// Create request
+	req, err := http.NewRequestWithContext(ctx, "PUT", uploadURL, bytes.NewReader(file.Data))
+	if err != nil {
+		return UploadFileResult{}, fmt.Errorf("failed to create upload request: %w", err)
+	}
+
+	// Set headers
+	req.Header.Set("Content-Type", file.Type)
+	for key, value := range requiredHeaders {
+		req.Header.Set(key, value)
+	}
+
+	// Execute request
+	resp, err := client.Do(req)
+	if err != nil {
+		return UploadFileResult{}, fmt.Errorf("file upload failed: %w", err)
+	}
+	defer func() {
+		err := resp.Body.Close()
```

Comment:
Could create a helper function for the deferred `resp.Body.Close()`, this happens in multiple places, including in `downloader.go`.